examples/go/tagging/gotag-fn/handler.go

```diff
@@ -0,0 +1,177 @@
+package function
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+
+	handler "github.com/openfaas-incubator/go-function-sdk"
+	"github.com/pelletier/go-toml"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/vapi/rest"
+	"github.com/vmware/govmomi/vapi/tags"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+// tagInfo is part of the toml vcconfig file
+type tagInfo struct {
+	Action string
+	Urn    string
+}
+
+// VCenter is part of the toml vcconfig file
+type vCenterAuth struct {
+	Password string
+	Server   string
+	User     string
+}
+
+// Config represents the toml vcconfig file
+type vcConfig struct {
+	Tag     tagInfo
+	VCenter vCenterAuth
+}
+
+// The request from vmware-event-router.
+type dataObj struct {
+	Data nest1 `json:"data"`
+}
+type nest1 struct {
+	Vm nest2 `json:"Vm"`
+}
+type nest2 struct {
+	Vm vmInfo `json:"Vm"`
+}
+type vmInfo struct {
+	Value string `json:"Value"`
+	Type  string `json:"Type"`
+}
+
+// Handle a function invocation
+func Handle(req handler.Request) (handler.Response, error) {
+	ctx := context.Background()
+
+	// Retrieve vCenter and tag info from the config file.
+	vcc := loadTomlCfg()
+
+	// Incoming request object.
+	var mor dataObj
+	err := json.Unmarshal(req.Body, &mor)
+	if err != nil {
+		return handler.Response{
+			Body:       nil,
+			StatusCode: http.StatusBadRequest,
+		}, err
+	}
+
+	if mor.Data.Vm.Vm.Value == "" || mor.Data.Vm.Vm.Type == "" {
+		return handler.Response{
+			Body:       nil,
+			StatusCode: http.StatusBadRequest,
+		}, errors.New("managed Reference Object information must not be empty")
+	}
+
+	// Fill information in the request into a govmomi type.
+	var moRef types.ManagedObjectReference
+	moRef.Value = mor.Data.Vm.Vm.Value
+	moRef.Type = mor.Data.Vm.Vm.Type
+
+	// Connect to vcenter with govmomi client
+	log.Println("Connecting to vCenter")
+	clients := connect(ctx, vcc.VCenter)
+	log.Println("Connection successful.")
+
+	// tag the VM moRef with tagID (URN)
+	clients.tagVM(ctx, moRef, vcc.Tag.Urn)
+
+	// Respond with a message that tagging of vm was complete
+	message := fmt.Sprintf(
+		"%v was tagged with %v",
+		moRef.Value,
+		vcc.Tag.Urn,
+	)
+	log.Println(string(message))
+
+	err = clients.Govmomi.Logout(ctx)
```

Comment:
perhaps it's better to extract the client logic into its own package and use a defer on an exported method to logout from all sessions (or persist them across invocation and only logout when SIGTERM/SIGINTERRUPT)

Reply:
Client logic has been extracted.

Reply:
Logout is only called with SIGTERM/SIGNINTERRUPT is sent.

Reply:
> Client logic has been extracted.

As per review comment below, consider making this part of this PR instead of a separate repo. There's some issues in the client code which could break this fn.

> Logout is only called with SIGTERM/SIGNINTERRUPT is sent.

Did you actually test that? Problem: if you use the same context and that is cancelled before calling `Logout` the govmomi call might fail with `context.Cancelled` and leak connections. Check the event router code for a workaround.

Reply:
I noticed `SIGTERM` in action last week, and I it does seem like, as you say, the context is cancelled before calling `Logout`. I thought at the time that maybe `Logout`, though not called in our situation, might be a backup path if the context failed to cancel. Is logout not necessary in our case? It doesn't seem like it if context is handling everything and session authentication is lost before we can properly log out. Here's what I saw when a pod shutdown (switching pods):

```
2020/03/13 23:05:58 stderr: 2020/03/13 23:05:58 got signal: terminated,
2020/03/13 23:05:58 stderr: log out of vSphere
2020/03/13 23:05:58 SIGTERM received.. shutting down server in 10s
2020/03/13 23:05:58 Removing lock-file : /tmp/.lock
2020/03/13 23:05:58 stderr: 2020/03/13 23:05:58 [entrypoint] SIGTERM received.. shutting down server in 10s
2020/03/13 23:05:58 stderr: 2020/03/13 23:05:58 vSphere logout failed: govmomi api logout failed: ServerFaultCode: The session is not authenticated.
```

Reply:
I'll definitely move the client code to this repo.